[hendo.park 박채운] Step3 코드리뷰 다시 요청드립니다.

hendo-park opened the pull request:
안녕하세요! 아까 말씀드린 것 처럼 서비스 레이어를 추가한 뒤에 코드를 수정해서 다시 풀리퀘스트를 보냅니다..!
보내주신 자료들이 큰 도움이 된 것 같습니다..!(코드도 그랬으면 좋겠네요..! ㅎㅎㅎ ㅠㅠㅋㅋㅋ)
잘부탁드리겠습니다.

Commit:
refactor : 초기화 과정 진행

step3 진행을 위한 초기화를 진행하였습니

Commit:
feat : jdbc, h2 데이터베이스 주입

build.gradle에 jdbc와 h2 데이터베이스 관련 dependency를 추가하고,
사용할 데이터베이스 schema를 설계하였습니다.

Commit:
feat : station 구현

station의 컨트롤러와 dao를 구현하였습니다.

Commit:
feat : Line 구현 (~step2)

Line의 컨트롤러와 Dao를 step2 부분까지 구현하였습니다.

Commit:
feat : Section 구현 (~step3)

Section을 구현하기 위한 Dao와 Controller을 구성하였습니다.
발생할 수 있는 에러들을 exceptions패키지로 분류하였습니다.
도메인 로직의 경우 sections에 존재하게 되며, 그에 맞는 테스트를 구상하였습니다.
sections를 구현하면서 Line의 경우 StationResponse를 반환할 수 있도록 약간의 리팩토링 및 구현을 진행하였습니다.

Commit:
refactor : distance를 RelativeDistance 일급 콜랙션으로 변경

요구사항에서 주어진 distance의 경우, 구간간의 거리를 나타내지만,
현재 방식에서 Section이 저장하는 정보는 상대적인 거리이기 때문에 명명을 변경할 필요가 있다고 생각하였습니다.
또한, 상대 거리의 계산 방식이 바뀔 수 있는 점을 고려해 일급객체로 나타내었습니다.
[질문]
이렇게 하다보니, VO class인 Section 객체와 쿼리문 간의 매핑이 정확히 일치하지 않고,
RelativeDistance getter메서드를 두개가지는 구현이 이루어졌는데, 옳지 못한 방법이었을까요?

Commit:
refactor : RequestMapping으로 중복된 부분을 꺼냄

RequestMapping으로 컨트롤러에서 중복된 구문들을 꺼냈습니다.

Commit:
refactor : StationService 구현

StationService를 구현함으로서, Controller와 Dao의 구분을 보다 명확히 하고,
생성시 이름 중복, 삭제시 id 미존재의 경우를 에러처리 하였습니다.

Commit:
refactor : LineService, SectionService 구현 및 전체적인 리팩토링

서비스 레이어를 생성하여서 컨트롤러와 dao 객체의 책임을 더 한정시키고,
전체적인 네이밍을 포함한 리팩토링을 진행하였습니다.

Commit:
refactor : section table 객체 외래키 설정

section 객체는 line_id 와 station_id를 외래키로 가지고 있기 때문에, 외래키로 설정하였습니다.
또한, 외래키가 지워지면 같이 삭제되도록 하였습니다.

Hue9010 commented on src/main/java/subway/section/SectionDao.java:
```diff
@@ -0,0 +1,56 @@
+
+    public List<Section> findAll() {
+        return jdbcTemplate.query("select * from section ", sectionMapper);
+    }
```
사용하지 않게 된 메서드(findAll)나 변수(Line 40 id)들은 그때그때 제거해주시는게 좋아요.
혹시나 나중에 필요하지 않을까 싶은 코드라면 git을 믿으시면 됩니다. 보통 다시 볼 일은 없습니다. 😋

Hue9010 commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,42 @@
+
+    public List<StationResponse> getStationResponses() {
+        return stationDao.findAll().stream()
+                .map(station -> new StationResponse(station.getId(), station.getName()))
```
Line 22에서 활용한 StationResponse.from을 활용하실 수 있을 것 같아요. 😉

Hue9010 commented on src/main/java/subway/section/RelativeDistance.java:
```diff
@@ -0,0 +1,34 @@
+package subway.section;
+
+public class RelativeDistance {
```
Value Object로 만드셨군요. 😮 👍

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    private void validateMakeStation(Long StationId, int distance) {
+        RelativeDistance StationRelativeDistance = getRelativeDistanceByStationId(StationId);
```
오타로 보이는데 변수의 시작은 소문자로! 😉

Hue9010 commented on src/main/java/subway/line/LineResponse.java:
```diff
@@ -18,6 +20,10 @@ public LineResponse(Long id, String name, String color, List<StationResponse> st
         this.stations = stations;
     }
 
+    public static LineResponse of(Line line, List<StationResponse> stations) {
```
from, of 상황에 맞춰 잘 사용하셨네요. 💯

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void createSection(Long lineId, SectionRequest sectionRequest) {
+        Sections sectionsByLineId = new Sections(sectionDao.findByLineId(lineId));
```
Sections의 변수명에 붙은 ByLineId는 불필요하지 않을까요? 🤔

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,60 @@
+        sectionDao.save(lineId,
+                sectionsByLineId.getExtendedStationId(upStationId,downStationId),
+                sectionsByLineId.calculateRelativeDistance(upStationId,downStationId,distance)
+        );
```
`upStationId`, `downStationId`, `distance` 이 세가지 데이터를 중심으로 로직을 수행하고 이 데이터를 저장하면서 데이터 중심적으로 구현이 되어 있습니다. 
이를 sectionRequest를 section으로 변환을 한다음 section을 가지고 로직을 수행하고 section을 저장하면 어떨까요? 
dao에 save, update는 upStationId 같은 값이 아니라 객체를 전달해주는 방법으로 변경해보세요.

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,60 @@
+    public void deleteSection(Long lineId, Long stationId) {
+        Sections sectionsByLineId = new Sections(sectionDao.findByLineId(lineId));
+        sectionsByLineId.validateDeleteStation(stationId);
+        sectionDao.deleteByLineIdAndStationId(lineId, stationId);
```
delete도 데이터 기준이 아닌 section 기준으로 제거해보세요. 
객체지향 프로그래밍을 학습하기 위한 차원으로 의도적으로 가능 한 모든 처리는 도메인 객체 기준으로 잡아보세요.
```java
public class SectionDao {
    public void delete(Section section);
```

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,60 @@
+
+@Service
+public class SectionService {
+    public static final int INITIAL_DISTANCE = 0;
```
상수라도 내부에서만 사용할 값이라면 private로!

Hue9010 commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,38 @@
+    private Long id;
+    private Long stationId;
+    private Long lineId;
+    private RelativeDistance relativeDistance;
```
거리보다는 위치의 느낌이 강하네요. 🤔
위치 정보를 의미하는게 맞다면 이건 역과 역 사이의 연속된 공간을 의미하는 구간이 아닌 역이 가져야 할 값으로 보여요.
이 위치를 잠깐 위경도로 치환해서 생각해보면 (38.6087, 127.0215)가 판교역의 위경도인건 문제없지만 판교역-이매역 구간의 위경도가 될 수 있을까요? 🤔

객체를 모델링 할 때는 그 객체한테 맞는 상태 값인지에 대한 고려도 해주셔야 해요. 객체를 모델링 하는 것이지 단순하게 관련된 데이터를 모아두는게 아니라는 점 조심해야 합니다! 😏

Hue9010 replied:
우선 너무 고민하지 마시고 이번 과제는 이 상태로 진행하죠. 바꾸지 마세요!! 😋

hendo-park replied:
라인상에서 기준역을 잡은뒤 그 기준역으로부터  역의 거리라고 보는것이 맞을거같은데, 개인적으로 역이 가지기에는 약간 애매한거같다는 생각도 드네요..! (아니면 제가 코드를 조금 애매하게 짠것일까요? 위경도랑은 조금 다른 개념이라고 생각되어요..!) 그치만 문제가 있다는 말씀에 대해서는 동의합니다..!! 일단 이부분에 대해서는 수정하지 않고 더 고민해보겠습니다!!

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,60 @@
+        int distance = lineRequest.getDistance();
+        sectionDao.save(lineId, upStationId, INITIAL_DISTANCE);
+        sectionDao.save(lineId, downStationId, distance);
+    }
```
내부에서만 사용한다면 private로!
접근 제어자는 기본적으로 private로 사용한다 생각하는게 좋습니다.

Hue9010 replied:
DTO(LineRequest)는 가능하다면 service layer에 도착하면 도메인으로 변환 후 사용하지 않는걸 추천드려요. DTO가 의미 그대로 데이터를 전송하기 위한 객체이기도 하지만 서로 다른 계층간 데이터를 전달하기 위한 객체입니다. 그런 면에서 service layer에서 dto를 주고 받는건 올바른 사용 측면이 아니라고 생각됩니다.

다만 DTO를 어느 수준까지 사용하는가에 대해서는 사람마다 조금씩 견해가 달라 참고만 해주시면 됩니다. 그럼에도 의견을 드리는 이유는 현재 보면 특정 값으로만 가능하겠다 싶은 부분들은 간단히 값을 이용해서 처리하고 계신걸로 보입니다. 그래서 그런 가능성을 줄일 수 있는 방안으로 DTO의 사용을 제한 적으로 해보는걸 요청 드리는 겁니다. 

가능한 선에서 service -> service 간 참조에서는 dto 사용을 제한 해보시는걸 추천 드립니다.

Hue9010 replied:
당장 생각 나는 방법으로는 `lineRequest.toSection()`, `lineRequest.toFirstSection()` 으로 section을 가져 올 수 있을 것 같아요.

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,62 @@
+
+    public LineResponse showLineById(Long lineId) {
+        lineFindValidate(lineId);
+        return makeLineResponseByLine(lineDao.findById(lineId));
```
lineDao.findById(lineId)을 두번 수행하네요. 두번 수행하기보다는 한번만 수행후 null check하는건 어떨까요? 🤔

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,127 @@
+        return sections.stream()
+                .map(Section::getStationId)
+                .anyMatch(lineId -> lineId.equals(stationId));
+    }
```
왠지 외부에서도 사용하다 바꼈을 것 같은데 지금은 private로 충분합니다!

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    public void validateSection(Long upStationId, Long downStationId, int distance) {
+        validateSameSection(upStationId, downStationId);
+        validateNoStations(upStationId, downStationId);
+        validateDistance(distance);
+        if (haveToMakeUpStation(upStationId, downStationId)) {
+            validateMakeStation(downStationId, upStationDistance(distance));
+        }
+        if (haveToMakeDownStation(upStationId, downStationId)) {
+            validateMakeStation(upStationId, downStationDistance(distance));
+        }
```
upStationId, downStationId에 해당하는 section을 처음 한번씩만 조회하면 매번 조회할 필요가 없어보여요.
성능에 문제가 있기 전엔 최적화에 대해서 크게 고려하실 필요가 없지만 약간은 과하게 sectin을 조회하는 걸로 보여 개선하는게 좋아 보입니다.

Hue9010 commented on src/main/java/subway/exceptions/ExceptionController.java:
```diff
@@ -0,0 +1,63 @@
+import subway.exceptions.sectionExceptions.SectionNoStationException;
+import subway.exceptions.sectionExceptions.SectionSameSectionException;
+
+@ControllerAdvice
+public class ExceptionController {
```
ControllerAdvice를 통한 예외 처리를 해주셨네요. 👍 👍

Hue9010 commented on src/main/java/subway/exceptions/ExceptionController.java:
```diff
@@ -0,0 +1,63 @@
+@ControllerAdvice
+public class ExceptionController {
+
+    @ExceptionHandler(LineNotFoundException.class)
+    public ResponseEntity lineNotFoundHandle() {
+        return ResponseEntity.badRequest().body("해당 노선을 찾을 수 없습니다");
+    }
```
예외 발생 시에 적절한 메시지를 넣어주고 해당 메시지를 활요 할 수도 있습니다. 😉 
```java
    @ExceptionHandler(LineNotFoundException.class)
    public ResponseEntity lineNotFoundHandle(LineNotFoundException e) {
        return ResponseEntity.badRequest().body(e.getMessage());
    }
```

Hue9010 commented on src/test/java/subway/section/SectionsTest.java:
```diff
@@ -0,0 +1,225 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+
+public class SectionsTest {
```
꼼꼼하게 테스트를 작성해주셔서 리뷰하는데 수월했네요. 👍
테스트할 것 은 적지만 
```java
public class Section {
    public int compareDistance(Section otherSection);
```
등등 다른 도메인도 테스트해보세요.

Hue9010 commented on src/test/java/subway/section/SectionsTest.java:
```diff
@@ -0,0 +1,225 @@
+        sections = new Sections(tempSections);
+    }
+
+    @DisplayName("스테이션 번호가 주어지면, 그 스테이션이 섹션 안에 있는지 확인한.")
+    @ParameterizedTest
+    @CsvSource({
+            "1,true", "4,false", "3,true", "10,false",
+    })
```
junit5 기능을 잘 써주셨네요 👍

Hue9010 reviewed (changes requested):
헨도! 이번 리뷰를 맡은 휴입니다. 😄
우선 잘 구현해주셨어요! 짧은 시간에 기존 pr에서 많은 것들이 바뀌고 개선 됐네요. 👍 
이전에 질문 주셨던건 아까 드린 내용으로 해결되신건지 궁금해요. 아니라면 질문주세요!
마찬가지로 피드백 보고 궁금하신 사항은 언제든 질문주시면 됩니다!

Commit:
refactor : 사용하지 않는 객체 삭제

사용하지 않는 객체를 삭제하였습니다.

Commit:
refactor : 에러메세지 처리, dto 사용제한

request객체들은 서비스레이어를 벗어나지 않도록 설정하고,
에러메세지들을 처리해주었습니다.

Commit:
refactor : validate 간소화

validate에서 발생하는 중복된 탐색을 간소화했습니다.

Commit:
refactor : 도메인에 관련된 Test를 구현

Commit:
refactor : SectionGenerateStrategy 패턴 구현

전략패턴을 통해 하행점 생성의 경우와 하행점 생성의 경우를 구현했습니다.
RelativeDistance 객체의 필요성을 그 이후 느끼지 못하여 삭제하였습니다.

Chaeoon-Park commented:
질문이있습니다..!

1. RelativeDistance관련 (코멘트에 단 부분)
라인상에서 기준역을 잡은뒤 그 기준역으로부터 역의 거리라고 보는것이 맞을거같은데, 개인적으로 역이 가지기에는 약간 애매한거같다는 생각도 드네요..! (아니면 제가 코드를 조금 애매하게 짠것일까요? 위경도랑은 조금 다른 개념이라고 생각되어요..!) 그치만 문제가 있다는 말씀에 대해서는 동의합니다..!! 일단 이부분에 대해서는 수정하지 않고 더 고민해보겠습니다!!

코멘트 이후 수정 --> 시간이 조금 있다고 판단되어서, 전략패턴으로 상행선과 하행선을 만들어 보다 명쾌하게(?) 보일수 있도록 수정하고자 노력하였고,
이름을 포지션으로 변경하였습니다. 혹시 어떻게 생각하시나요?

2. 네이밍문제
제가 끔찍한 작명 실력을 가지고 있는것인지... 네이밍때문에 코드가 길어지는듯한 느낌을 자꾸 받게 되는 것 같습니다 ㅠㅠ
그렇다고 변수명을 더 고민하고 있자니 시간이 너무가는거같기도 하고, 저만의 규약을 둬야하는걸까요? (예를들어 10자 이하 뭐 이런식으로요?)


마지막으로 혹시 피드백이 제대로 반영되지 않은 부분이 있다면 말씀해주시면 감사하겠습니다..!
혹시 제 이해가 부족하거나 잘못되었을까봐 걱정되네요..! ㅎㅎ ㅠㅠ

Chaeoon-Park commented:
아 하나만 더 말씀드리자면, sections 클래스에 validateAndGenerateStrategy라는 메서드가있는데, 이름에서 보다싶이 2가지 일을 하고있습니다. 하지만 전략의 생성과 검사는 한번에 이루어질만한 것? 이라고 생각되어서 (분리하면 비효율적으로 분리되는 경우 밖에 생각나지 않습니다.) 이렇게 제작하였는데 이정도는 관용적으로(?) 용납할 수 있는 메서드일까요..?

Commit:
refactor : 쿼리문 간편화 및 dto사용 놓친 부분 리팩토링

Commit:
refactor : section 패키지 리팩토링 (쿼리문 변경에 따른)

SectionDao 의 findByLindId 쿼리를 orderby로 처리하고, 그에 따라 메서드들을 변경하고 리팩토링했습니다.
SectionIllegalDistanceException 처리의 경우 책임을 전략에서 맡도록 변경하였습니다.

Commit:
refactor : 패키지 분리

패키지를 분리하였습니다.

Chaeoon-Park commented:
질문 이후에 리팩토링을 조금 더 진행하였습니다..!!!
리팩토링을 크게(?) 진행하였긴 하지만 위의 질문은 거의 동일한 것 같습니다!
